pkg/code/data/nonce/postgres/model.go

```diff
@@ -162,48 +171,77 @@ func dbGetAllByState(ctx context.Context, db *sqlx.DB, state nonce.State, cursor
-// todo: We may need to tune the offset based on pool size and environment, but it
-// should be sufficiently good enough for now.
+// We query a random nonce by first selecting any available candidate from the
+// total set, applying an upper limit of 100, and _then_ randomly shuffling the
+// results and selecting the first. By bounding the size before ORDER BY random(),
+// we avoid having to shuffle large sets of results.
+//
+// Previously, we would use OFFSET FLOOR(RANDOM() * 100). However, if the pool
+// (post filter) size was less than 100, any selection > pool size would result
+// in the OFFSET being set to zero. This meant random() disappeared for a subset
+// of values. In practice, this would result in a bias, and increased contention.
+//
+// For example, 50 Available nonce's, 25 Claimed (expired), 25 Reserved. With Offset:
+//
+//  1. 50% of the time would be a random Available.
+//  2. 25% of the time would be a random expired Claimed.
+//  3. 25% of the time would be _the first_ Available.
+//
+// This meant that 25% of the time would not be random. As we pull from the pool,
+// this % only increases, further causing contention.
+//
+// Performance wise, this approach is slightly worse, but the vast majority of the
+// time is spent on the scan and filter. Below are two example query plans (from a
+// small dataset in an online editor).
+//
+// QUERY PLAN (OFFSET):
+//
+//	Limit  (cost=17.80..35.60 rows=1 width=140) (actual time=0.019..0.019 rows=0 loops=1)
+//	->  Seq Scan on codewallet__core_nonce  (cost=0.00..17.80 rows=1 width=140) (actual time=0.016..0.017 rows=0 loops=1)
+//	      Filter: ((signature IS NOT NULL) AND (purpose = 1) AND ((state = 0) OR ((state = 2) AND (claim_expires_at < 200))))
+//	      Rows Removed by Filter: 100
+//
+// Planning Time: 0.046 ms
+// Execution Time: 0.031 ms
+//
+// QUERY PLAN (ORDER BY):
+//
+//	Limit  (cost=17.82..17.83 rows=1 width=148) (actual time=0.018..0.019 rows=0 loops=1)
+//	->  Sort  (cost=17.82..17.83 rows=1 width=148) (actual time=0.018..0.018 rows=0 loops=1)
+//	      Sort Key: (random())
+//	      Sort Method: quicksort  Memory: 25kB
+//	      ->  Subquery Scan on sub  (cost=0.00..17.81 rows=1 width=148) (actual time=0.015..0.016 rows=0 loops=1)
+//	            ->  Limit  (cost=0.00..17.80 rows=1 width=140) (actual time=0.015..0.015 rows=0 loops=1)
+//	                  ->  Seq Scan on codewallet__core_nonce  (cost=0.00..17.80 rows=1 width=140) (actual time=0.015..0.015 rows=0 loops=1)
+//	                        Filter: ((signature IS NOT NULL) AND (purpose = 1) AND ((state = 0) OR ((state = 2) AND (claim_expires_at < 200))))
+//	                        Rows Removed by Filter: 100
+//
+// Planning Time: 0.068 ms
+// Execution Time: 0.037 ms
+//
+// Overall, the Seq Scan and Filter is the bulk of the work, with the ORDER BY RANDOM()
+// adding a small (fixed) amount of overhead. The trade-off is negligible time complexity
+// for more reliable semantics.
```

Comment:
This is probably a thing to really double check.

1. I discovered this oddity when testing the distributions (which was also highlighted in previous comments). I _think_ this change is likely worth it given the minuscule (theoretical) performance difference.
2. The previous approach had a fallback query, which I _think_ was to address the `OFFSET` selecting an offset out of range, but in testing the `OFFSET` was just set to 0. The new approach removes this issue (but if it was for another issue, that's important to know!)